rust/nasl-builtin-network/src/socket.rs

```diff
@@ -0,0 +1,731 @@
+use pkcs8::der::Decode;
+use rustls::{
+    pki_types::{PrivateKeyDer, PrivatePkcs8KeyDer, ServerName},
+    RootCertStore,
+};
+
+use crate::{get_kb_item, get_pos_port, OpenvasEncaps};
+
+const MTU: usize = 512 - 60 - 8;
+
+type NaslSocketFunction =
+    fn(&NaslSockets, &Register, &Context) -> Result<NaslValue, FunctionErrorKind>;
+
+pub struct Interval {
+    interval: Duration,
+    last_tick: SystemTime,
+}
+
+impl Interval {
+    pub fn tick(&mut self) {
+        if let Ok(since) = SystemTime::now().duration_since(self.last_tick) {
+            if since < self.interval {
+                sleep(self.interval - since);
+            }
+        }
+        self.last_tick = SystemTime::now();
+    }
+}
+
+struct TLSConfig {
+    config: rustls::ClientConfig,
+    server: ServerName<'static>,
+}
+
+/// A TLS Connection consists of the underlying socket, an optional TLS Connection and an optional buffer
+struct TCPConnection {
+    socket: TcpStream,
+    _tls_connection: Option<rustls::ClientConnection>,
+    _buffer: Option<Vec<u8>>,
+}
+
+struct UDPConnection {
+    socket: UdpSocket,
+    buffer: Vec<u8>,
+}
+
+enum NaslSocket {
+    // The TCP Connection is boxed, because it uses allot of space
+    // This way the size of the enum is reduced
+    Tcp(Box<TCPConnection>),
+    Udp(UDPConnection),
+    Close,
+}
+
+#[derive(Default)]
+struct Handles {
+    handles: Vec<NaslSocket>,
+    closed_fd: Vec<usize>,
+}
+
+#[derive(Default)]
+pub struct NaslSockets {
+    handles: RwLock<Handles>,
+    interval: Option<RwLock<Interval>>,
+}
+
+impl NaslSockets {
+    fn open_udp(addr: &str, port: u16) -> Result<NaslSocket, FunctionErrorKind> {
+        let socket = UdpSocket::bind("0.0.0.0:0")?;
+        socket.connect(format!("{addr}:{port}"))?;
+        socket.set_read_timeout(Some(Duration::from_secs(1)))?;
+        Ok(NaslSocket::Udp(UDPConnection {
+            socket,
+            buffer: vec![],
+        }))
+    }
+
+    fn open_tcp(
+        ip: &str,
+        port: u16,
+        bufsz: Option<i64>,
+        timeout: Option<Duration>,
+        tls_config: Option<&TLSConfig>,
+    ) -> Result<NaslSocket, FunctionErrorKind> {
+        // Resolve Address and Port to SocketAddr
+        let sock = format!("{ip}:{port}").to_socket_addrs()?.next().ok_or(
+            FunctionErrorKind::WrongArgument(format!(
+                "the given address and port do not correspond to a valid address: {ip}:{port}"
+            )),
+        )?;
+        // Create Vec depending of buffer size
+        let buffer = if let Some(bufsz) = bufsz {
+            if bufsz > 0 {
+                Some(Vec::with_capacity(bufsz as usize))
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+
+        // Create socket for connection
+        let socket = if let Some(timeout) = timeout {
+            TcpStream::connect_timeout(&sock, timeout)?
+        } else {
+            TcpStream::connect(sock)?
```

Comment:
probably also done so in the OG implementation: should we work without a timeout at all?

Reply:
No sure, if they are needed. They are also really messy...
In the c implementation: if the given timeout is negative, it is set to 10, if it was not given, it is set to the double of the globally defined recv timeout. But I also know about services with a "sleep mode" when no one is connected for a specified time and wakes up on an incoming connection. But this normally leads to the first connection try to fail. In such cases a timeout could make sense. On the other hand it is not used in the feed, so it will always be the recv timeout times 2...

Reply:
I will set the default timeout to 10 seconds for now